File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireAgentFrontmatterRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that requires agent files to have proper frontmatter.
 */
public enum RequireAgentFrontmatterRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "requireAgentFrontmatter";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        Path directory = root.resolve(".claude/agents");
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        if (!HarnessCheckHelper.isSafeDirectory(root, directory)) {
            return List.of(new Finding(severity, CATEGORY, ".claude/agents must contain at least one .md agent"));
        }
        try (Stream<Path> stream = Files.list(directory)) {
            List<Path> files = stream
                    .filter(f -> !Files.isSymbolicLink(f) && f.getFileName().toString().endsWith(".md"))
                    .toList();
            if (files.isEmpty()) {
                return List.of(new Finding(severity, CATEGORY, ".claude/agents must contain at least one .md agent"));
            }
            Pattern namePattern = Pattern.compile("(?m)^name:\\s*[-a-z0-9]+\\s*$");
            Pattern descPattern = Pattern.compile("(?m)^description:\\s*.+$");
            return files.stream()
                    .flatMap(f -> validateAgentFile(root, f, namePattern, descPattern, severity).stream())
                    .toList();
        } catch (IOException e) {
            return List.of();
        }
    }

    private List<Finding> validateAgentFile(Path root, Path file, Pattern namePattern, Pattern descPattern, String severity) throws MojoExecutionException {
        String text = HarnessCheckHelper.readFile(root, file);
        String relative = root.relativize(file).toString();
        return Stream.of(
                text.startsWith("---") ? Stream.empty() : Stream.of(new Finding(severity, CATEGORY, "agent missing frontmatter: " + relative)),
                namePattern.matcher(text).find() ? Stream.empty() : Stream.of(new Finding(severity, CATEGORY, "agent missing name: " + relative)),
                descPattern.matcher(text).find() ? Stream.empty() : Stream.of(new Finding(severity, CATEGORY, "agent missing description: " + relative))
        ).flatMap(s -> s).toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/RequireAgentFrontmatterRule.kt
package ai.harness.gradle.rules

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.extension
import kotlin.io.path.isDirectory
import kotlin.io.path.isRegularFile
import kotlin.io.path.listDirectoryEntries
import kotlin.io.path.readText
import kotlin.io.path.relativeTo
import java.io.File
import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessCheckRule
import ai.harness.gradle.HarnessPsiResults

/**
 * Rule that requires agents to have proper frontmatter.
 */
object RequireAgentFrontmatterRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireAgentFrontmatter"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> {
		val category = "requireAgentFrontmatter"
		val severity = HarnessCheck.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject
		val parametersObj = catObj?.get("parameters")?.jsonObject
		val messagesObj = catObj?.get("messages")?.jsonObject
		return if (catObj == null || parametersObj == null || messagesObj == null) {
			emptyList()
		} else {
			val directory = HarnessCheck.stringFrom(parametersObj, "directory")
			val dirPath = root / directory
			if (!dirPath.isDirectory()) {
				val msg = HarnessCheck.stringFrom(messagesObj, "missingDirectory").takeIf { it.isNotEmpty() } ?: ".claude/agents must contain at least one .md agent"
				listOf(Finding(severity, category, msg))
			} else {
				val files = try {
					dirPath.listDirectoryEntries().filter { it.isRegularFile() && it.extension == "md" }
				} catch (_: Exception) {
					emptyList()
				}
				if (files.isEmpty()) {
					val msg = HarnessCheck.stringFrom(messagesObj, "missingAgent").takeIf { it.isNotEmpty() } ?: ".claude/agents must contain at least one .md agent"
					listOf(Finding(severity, category, msg))
				} else {
					files.flatMap { file ->
						val text = file.readText()
						if (!text.startsWith("---")) {
							val msg = HarnessCheck.stringFrom(messagesObj, "missingFrontmatter").takeIf { it.isNotEmpty() } ?: "agent missing frontmatter: ${file.relativeTo(root)}"
							listOf(Finding(severity, category, msg))
						} else {
							listOfNotNull(
								if (!"""(?m)^name:\s*[-a-z0-9]+\s*$""".toRegex().containsMatchIn(text)) {
									val msg = HarnessCheck.stringFrom(messagesObj, "missingName").takeIf { it.isNotEmpty() } ?: "agent missing name: ${file.relativeTo(root)}"
									Finding(severity, category, msg)
								} else null,
								if (!"""(?m)^description:\s*.+$""".toRegex().containsMatchIn(text)) {
									val msg = HarnessCheck.stringFrom(messagesObj, "missingDescription").takeIf { it.isNotEmpty() } ?: "agent missing description: ${file.relativeTo(root)}"
									Finding(severity, category, msg)
								} else null
							)
						}
					}
				}
			}
		}
	}
}
